eventstore/test_whatsapp_actions.py

```diff
@@ -244,6 +245,61 @@ def test_delivery_failed_error(self):
             handle_event(event)
             h.assert_called_once_with(event)
 
+    @override_settings(RAPIDPRO_OPTOUT_FLOW="test-flow-uuid")
+    def test_fallback_channel_delivery_failure_error(self):
+        """
+        If the event is of type Failed, and uses the fallback channel,
+        then it should trigger the message delivery failed action
+        """
+        event = Event.objects.create()
+        event.fallback_channel = True
+        event.recipient_id = "27820001001"
+        event.timestamp = datetime.datetime(2018, 2, 15, 11, 38, 20, tzinfo=UTC)
+
+        DeliveryFailure.objects.create(number_of_failures=5, contact_id="27820001001")
+
+        with patch("eventstore.tasks.rapidpro") as p:
+            handle_fallback_delivery_error(event)
+
+        p.create_flow_start.assert_called_once_with(
+            extra={"optout_reason": "sms_failure", "timestamp": 1518694700},
+            flow="test-flow-uuid",
+            urns=["whatsapp:27820001001"],
+        )
+
+    def test_fallback_channel_delivery_failure_less_than_5(self):
+        """
+        If the event is of type Failed, and uses the fallback channel,
+        but delivery failures are less than 5, should not call
+        the rapidpro flow
+        """
+        event = Event.objects.create()
+        event.fallback_channel = True
+        event.recipient_id = "27820001001"
+        event.timestamp = datetime.datetime(2018, 2, 15, 11, 38, 20, tzinfo=UTC)
+
+        with patch("eventstore.tasks.rapidpro") as p:
+            handle_fallback_delivery_error(event)
+
+        p.create_flow_start.assert_not_called()
+        df = DeliveryFailure.objects.get(contact_id="27820001001")
+        self.assertEqual(df.number_of_failures, 1)
+
+    def test_fallback_channel_successful_delivery_event(self):
+        """
+        If the event uses the fallback channel, but is a successful delivery,
+        should not call the rapidpro flow
+        """
+        event = Event.objects.create()
+        event.fallback_channel = True
+        event.recipient_id = "27820001001"
+        event.timestamp = datetime.datetime(2018, 2, 15, 11, 38, 20, tzinfo=UTC)
+
+        with patch("eventstore.tasks.rapidpro") as p:
+            handle_fallback_delivery_error(event)
+
+        p.create_flow_start.assert_not_called()
```

Comment:
This should not only not call the flow, but should also reset the `number_of_failures` to `0`